src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -533,7 +533,16 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedBlock(tree: untpd.Block, pt: Type)(implicit ctx: Context) = track("typedBlock") {
     val exprCtx = index(tree.stats)
     val stats1 = typedStats(tree.stats, ctx.owner)
-    val expr1 = typedExpr(tree.expr, pt)(exprCtx)
+
+    // A block might appear in function position after desugaring (see
+    // i1503.scala), so adaptation may insert a `.apply` at the end of expr1.
+    // When this happens we simply remove the select to make expr1 a valid
+    // block expression. The `.apply` will be added back around the block when
+    // the block itself is adapted later.
+    val expr1 = typedExpr(tree.expr, pt)(exprCtx) match {
```

Comment:
But what if the .apply was added explicitly by the user? Then it should not be removed. It seems we need another solution.